target/linux/ipq40xx/files/drivers/net/phy/qca807x.c

```diff
@@ -0,0 +1,834 @@
+		case QCA807X_FC_MDI_CROSSOVER_AUTO:
+			phydev->mdix_ctrl = ETH_TP_MDI_AUTO;
+			break;
+		}
+	}
+
+	if (phydev->autoneg == AUTONEG_ENABLE && phydev->autoneg_complete)
+		phy_resolve_aneg_pause(phydev);
+
+	return 0;
+}
+
+static int qca807x_read_fiber_status(struct phy_device *phydev, bool combo_port)
+{
+	int ss, err, page, lpa, old_link = phydev->link;
+
+	/* Only combo port has dual pages */
+	if (combo_port) {
+		/* Check whether fiber page is set and set if needed */
+		page = phy_read(phydev, QCA807X_CHIP_CONFIGURATION);
+		if (page & QCA807X_BT_BX_REG_SEL) {
+			page &= ~QCA807X_BT_BX_REG_SEL;
+			phy_write(phydev, QCA807X_CHIP_CONFIGURATION, page);
+		}
+	}
+
+	/* Update the link, but return if there was an error */
+	err = genphy_update_link(phydev);
+	if (err)
+		return err;
+
+	/* why bother the PHY if nothing can have changed */
+	if (phydev->autoneg == AUTONEG_ENABLE && old_link && phydev->link)
+		return 0;
+
+	phydev->speed = SPEED_UNKNOWN;
+	phydev->duplex = DUPLEX_UNKNOWN;
+	phydev->pause = 0;
+	phydev->asym_pause = 0;
+
+	if (phydev->autoneg == AUTONEG_ENABLE && phydev->autoneg_complete) {
+		lpa = phy_read(phydev, MII_LPA);
+		if (lpa < 0)
+			return lpa;
+
+		linkmode_mod_bit(ETHTOOL_LINK_MODE_Autoneg_BIT,
+				 phydev->lp_advertising, lpa & LPA_LPACK);
+		linkmode_mod_bit(ETHTOOL_LINK_MODE_1000baseX_Full_BIT,
+				 phydev->lp_advertising, lpa & LPA_1000XFULL);
+		linkmode_mod_bit(ETHTOOL_LINK_MODE_Pause_BIT,
+				 phydev->lp_advertising, lpa & LPA_1000XPAUSE);
+		linkmode_mod_bit(ETHTOOL_LINK_MODE_Asym_Pause_BIT,
+				 phydev->lp_advertising,
+				 lpa & LPA_1000XPAUSE_ASYM);
+
+		phy_resolve_aneg_linkmode(phydev);
+	}
+
+	/* Read the QCA807x PHY-Specific Status register fiber page,
+	 * which indicates the speed and duplex that the PHY is actually
+	 * using, irrespective of whether we are in autoneg mode or not.
+	 */
+	ss = phy_read(phydev, QCA807X_PHY_SPECIFIC_STATUS);
+	if (ss < 0)
+		return ss;
+
+	if (ss & QCA807X_SS_SPEED_AND_DUPLEX_RESOLVED) {
+		switch (FIELD_GET(QCA807X_SS_SPEED_MASK, ss)) {
+		case QCA807X_SS_SPEED_10:
+			phydev->speed = SPEED_10;
+			break;
+		case QCA807X_SS_SPEED_100:
+			phydev->speed = SPEED_100;
+			break;
+		case QCA807X_SS_SPEED_1000:
+			phydev->speed = SPEED_1000;
+			break;
+		}
+
+		if (ss & QCA807X_SS_DUPLEX)
+			phydev->duplex = DUPLEX_FULL;
+		else
+			phydev->duplex = DUPLEX_HALF;
+	}
+
+	return 0;
+}
+
+static int qca807x_read_status(struct phy_device *phydev)
+{
+	int val;
+
+	/* Check for Combo port */
+	if (phy_read(phydev, QCA807X_CHIP_CONFIGURATION)) {
+		/* Check for fiber mode first */
+		if (linkmode_test_bit(ETHTOOL_LINK_MODE_FIBRE_BIT, phydev->supported)) {
+			/* Check for actual detected media */
+			val = phy_read(phydev, QCA807X_MEDIA_SELECT_STATUS);
+			if (val & QCA807X_MEDIA_DETECTED_COPPER) {
+				qca807x_read_copper_status(phydev, true);
+			} else if ((val & QCA807X_MEDIA_DETECTED_1000_BASE_X) ||
+				   (val & QCA807X_MEDIA_DETECTED_100_BASE_FX)) {
```

Comment:
This seems to exclude speed 10 which you still test for on line 545 above: is that intentional?

Reply:
Yes, as this tests whether 1000Base-X or 100Base-FX fiber is detected, there is no 10Base fiber support.
But I left the test above as the PHY specific register has the same bits for speed status in the fiber page as in the copper one.

Reply:
OK, but it doesn't seem to make sense to have code testing for unsupported situations?

Reply:
It has the bits for speed 10 documented in the datasheet, so why not have it?

Reply:
Depends whether there's hardware that actually support this. Otherwise you're just wasting space with code that will never be executed IMHO.

Reply:
Actually it would make more sense to have this if both copper/fiber were a single function, since they share quite a bit of code. But I can see reasons to keep them separated to.

Reply:
> Depends whether there's hardware that actually support this. Otherwise you're just wasting (a bit of) space with code that will never be executed IMHO.

Well, since we already have the "implementation" now, I would not actively make the driver "worse" by removing it again, unless we talk about huge amounts or substantially complex code.

Reply:
@adschm I'm not sure I follow? My point is there's a contradiction between two code paths, one which exclude the other from ever happening. Either there is hardware that supports 10, and the case must be treated in both paths, or there isn't, and it seems odd to keep dead code. My 2c though.

Reply:
I mean, I can drop the case, it's only 2 lines though.
I intentionally made fiber and copper separate as fiber requires Clause 37 resolving of LPA while copper can use the generic one.
It also has a different PHY specific register page and this way it's much easier to follow the code and not get lost in a bunch of if-s.

Reply:
Dropped the 10M check

Reply:
Sorry for being pedantic 😉

Reply:
Nah, I am glad that somebody took a look at the code.
Saves me the pain with the mainteiners.